src/Microsoft.ML.SearchSpace/Parameter.cs

```diff
@@ -0,0 +1,278 @@
+            {
+                new StringEnumConverter(),
+            },
+        };
+
+        private readonly object _value;
+
+        private Parameter(object value, ParameterType type)
+        {
+            this._value = value;
+            this.ParameterType = type;
+        }
+
+        public static Parameter FromDouble(double value)
+        {
+            return new Parameter(value, ParameterType.Float);
+        }
+
+        public static Parameter FromFloat(float value)
+        {
+            return new Parameter(value, ParameterType.Float);
+        }
+
+        public static Parameter FromLong(long value)
+        {
+            return new Parameter(value, ParameterType.Integer);
+        }
+
+        public static Parameter FromInt(int value)
+        {
+            return new Parameter(value, ParameterType.Integer);
+        }
+
+        public static Parameter FromString(string value)
+        {
+            return new Parameter(value, ParameterType.Integer);
+        }
+
+        public static Parameter FromBool(bool value)
+        {
+            return new Parameter(value, ParameterType.Bool);
+        }
+
+        public static Parameter FromEnum<T>(T value) where T : struct, Enum
+        {
+            return Parameter.FromEnum(value, typeof(T));
+        }
+
+        public static Parameter FromIEnumerable<T>(IEnumerable<T> values)
+        {
+            // check T
+            return Parameter.FromIEnumerable(values as IEnumerable);
+        }
+
+        private static Parameter FromIEnumerable(IEnumerable values)
+        {
+            return new Parameter(values, ParameterType.Array);
+        }
+
+        private static Parameter FromEnum(Enum e, Type t)
+        {
+            return Parameter.FromString(Enum.GetName(t, e));
+        }
+
+        public static Parameter FromObject<T>(T value) where T : class
+        {
+            return Parameter.FromObject(value, typeof(T));
+        }
+
+        private static Parameter FromObject(object value, Type type)
+        {
+            var param = value switch
+            {
+                int i => Parameter.FromInt(i),
+                long l => Parameter.FromLong(l),
+                double d => Parameter.FromDouble(d),
+                float f => Parameter.FromFloat(f),
+                string s => Parameter.FromString(s),
+                bool b => Parameter.FromBool(b),
+                IEnumerable vs => Parameter.FromIEnumerable(vs),
+                Enum e => Parameter.FromEnum(e, e.GetType()),
+                _ => null,
+            };
+
+            if (param != null)
+            {
+                return param;
+            }
+            else
+            {
+                var parameter = Parameter.CreateNestedParameter();
+                var properties = type.GetProperties(System.Reflection.BindingFlags.Public | System.Reflection.BindingFlags.Instance)
+                        .Where(p => p.CanRead && p.CanWrite);
```

Comment:
Do you need to check for writing here? Shouldn't you be able to use `public readonly` members? If not, you should document that somewhere.

Reply:
`readonly` Property won't be deserialized by Json.Net doesn't it

Reply:
There are ways to do it if needed. Does that hold true if you aren't using Newtonsoft anymore?

Reply:
### Resolved